test/FunctionalTests/Microsoft.OData.Core.Tests/Json/PooledByteBufferWriterTests.cs

```diff
@@ -0,0 +1,187 @@
+
+            bufferWriter.Advance(200);
+            Assert.Equal(56, bufferWriter.FreeCapacity);
+            Assert.Equal(200, bufferWriter.WrittenCount);
+            Assert.Equal(256, bufferWriter.Capacity);
+
+            buffer = bufferWriter.GetMemory(200);
+            buffer.Slice(0, 200).Span.Fill(1);
+            bufferWriter.Advance(200);
+
+            Assert.Equal(112, bufferWriter.FreeCapacity);
+            Assert.Equal(400, bufferWriter.WrittenCount);
+            Assert.Equal(512, bufferWriter.Capacity);
+        }
+
+        [Fact]
+        public void GetSpan_NewBufferDoNotOverwritePreviousOne()
+        {
+            using PooledByteBufferWriter bufferWriter = new PooledByteBufferWriter(256);
+            Span<byte> buffer = bufferWriter.GetSpan(1);
+            buffer[0] = 143;
+
+            bufferWriter.Advance(1);
+
+            buffer = bufferWriter.GetSpan(1);
+            Assert.NotEqual(143, buffer[0]);
+        }
+
+        [Fact]
+        public void GetMemory_NewBufferDoNotOverwritePreviousOne()
+        {
+            using PooledByteBufferWriter bufferWriter = new PooledByteBufferWriter(256);
+            Memory<byte> buffer = bufferWriter.GetMemory(1);
+            buffer.Span[0] = 143;
+
+            bufferWriter.Advance(1);
+
+            buffer = bufferWriter.GetMemory(1);
+            Assert.NotEqual(143, buffer.Span[0]);
+        }
+
+        [Fact]
+        public void Clear_ResetsTheBuffer()
+        {
+            using PooledByteBufferWriter bufferWriter = new PooledByteBufferWriter(256);
+            Span<byte> buffer = bufferWriter.GetSpan(200);
+            bufferWriter.Advance(200);
+
+            Assert.Equal(200, bufferWriter.WrittenCount);
+            Assert.Equal(56, bufferWriter.FreeCapacity);
+            Assert.Equal(256, bufferWriter.Capacity);
+
+            bufferWriter.Clear();
+
+            Assert.Equal(0, bufferWriter.WrittenCount);
+            Assert.Equal(256, bufferWriter.FreeCapacity);
+            Assert.Equal(256, bufferWriter.Capacity);
+            Assert.Equal(0, bufferWriter.WrittenMemory.Length);
+        }
+
+        [Fact]
+        public void WrittenMemory_ReturnsDataWrittenSoFar()
+        {
+            using PooledByteBufferWriter bufferWriter = new PooledByteBufferWriter(256);
+            Span<byte> buffer = bufferWriter.GetSpan(5);
+            for (int i = 0; i < 5; i++)
+            {
+                buffer[i] = (byte)(i + 1);
+            }
+
+            bufferWriter.Advance(5);
+
+            buffer = bufferWriter.GetSpan(5);
+            for (int i = 0; i < 5; i++)
+            {
+                buffer[i] = (byte)(i + 6);
+            }
+
+            bufferWriter.Advance(5);
+
+            ReadOnlySpan<byte> data = bufferWriter.WrittenMemory.Span;
+            Assert.Equal(10, data.Length);
+            for (int i = 0; i < data.Length; i++)
+            {
+                Assert.Equal(i + 1, data[i]);
+            }
+        }
+    }
+}
+
+#endif
```

Comment:
One test that you do not have, and might not be safe to have, is one where you hit the 1GB mark and grow the array to `MaximumBufferSize`. I don't know if it'd be possible to use reflection to modify the `MaximumBufferSize` to a much lower value (in a test) to verify that the relevant logic works fine.

Reply:
https://github.com/OData/odata.net/pull/2792/files#r1396784482
This would also help you write a test to verify the behavior (exception thrown) when the `MaximumBufferSize` limit is breached.

Reply:
I have added a test to verify the exception is thrown when the max buffer size is reached. But I didn't add a test to show that the buffer grows to the max size when the max size/2 threshold is reached. I can't use reflection to change the max buffer size because it's a `const`, i.e. set at compile time, and not a `readonly static` field. So I would actually have to allocate a 1GB+ array to test this and I'm not sure that's safe to do in a test and whether it'll run successfully on the CI environment,